File: bank-kyc-system/src/main/java/com/bank/project/controller/CustomerController.java
package com.bank.project.controller;

import java.io.IOException;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.ModelAttribute;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import com.bank.project.dto.AccountInfoDTO;
import com.bank.project.dto.KycUploadRequestDTO;
import com.bank.project.dto.UserRegisterDTO;
import com.bank.project.entity.KycApplication;
import com.bank.project.enums.Role;
import com.bank.project.service.CustomerService;

import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;


@RestController
@RequestMapping("/api/customer")
@PreAuthorize("hasRole('CUSTOMER')")
@RequiredArgsConstructor
public class CustomerController {

    private final CustomerService customerService ;
   
    @PostMapping("/register-customer")
    public ResponseEntity<String> registerCustomer(@RequestBody @Valid UserRegisterDTO dto) {
        dto.setRole(Role.CUSTOMER); 
        customerService.registerCustomer(dto);
        return ResponseEntity.status(HttpStatus.CREATED).body("Customer registered successfully");
    }

//    @PostMapping("/upload-kyc")
//    public ResponseEntity<String> uploadKyc(
//            @RequestParam Long userId,
//            @ModelAttribute KycUploadRequestDTO dto) throws IllegalStateException, IOException {
//        
//        customerService.uploadKycDocuments(userId, dto);
//        return ResponseEntity.ok("KYC documents uploaded successfully");
//    }
   
    @PostMapping("/upload-kyc")
    public ResponseEntity<String> uploadKyc(@ModelAttribute KycUploadRequestDTO dto) throws IOException {
        customerService.uploadKycDocumentsForCurrentUser(dto);
        return ResponseEntity.ok("KYC documents uploaded successfully");
    }

//    @GetMapping("/account")
//    public ResponseEntity<AccountInfoDTO> getAccountInfo(@RequestHeader("email") String email) {
//        AccountInfoDTO dto = customerService.getAccountInfo(email);
//        return ResponseEntity.ok(dto);
//    }
//    

    @GetMapping("/account")
    public ResponseEntity<AccountInfoDTO> getAccountInfo(Authentication authentication) {
        String email = authentication.getName();
        AccountInfoDTO dto = customerService.getAccountInfoByEmail(email); // ✅ corrected
        return ResponseEntity.ok(dto);
    }


    @GetMapping("/account/{userId}")
    public ResponseEntity<AccountInfoDTO> getAccountInfoByUserId(@PathVariable Long userId) {
        AccountInfoDTO dto = customerService.getAccountInfoByUserId(userId);
        return ResponseEntity.ok(dto);
    }

 
    
    @PreAuthorize("hasRole('CUSTOMER')")
    @GetMapping("/self")
    public ResponseEntity<KycApplication> getOwnKyc() {
        return ResponseEntity.ok(customerService.getKycForCurrentUser());
    }

   

}

File: BankKycSystemdotnet/BankKycSystem/Controllers/CustomerController.cs
using BankKycSystem.DTOs;
using BankKycSystem.Entities;
using BankKycSystem.Services.Interfaces;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using System.Security.Claims;

namespace BankKycSystem.Controllers
{
    [ApiController]
    [Route("api/customer")]
    public class CustomerController : ControllerBase
    {
        private readonly ICustomerService _customerService;

        public CustomerController(ICustomerService customerService)
        {
            _customerService = customerService;
        }

        [HttpPost("register-customer")]
        public async Task<IActionResult> RegisterCustomer([FromBody] UserRegisterDTO dto)
        {
            await _customerService.RegisterCustomerAsync(dto);
            return Created("", "Customer registered successfully");
        }
        [Authorize]
        [Authorize(Roles = "CUSTOMER")]
        [HttpPost("upload-kyc")]
        public async Task<IActionResult> UploadKyc([FromForm] KycUploadRequestDTO dto)
        {
            var email = User.FindFirstValue(ClaimTypes.Name);
            await _customerService.UploadKycDocumentsForCurrentUserAsync(dto, email);
            return Ok("KYC documents uploaded successfully");
        }

        [Authorize(Roles = "CUSTOMER")]
        [HttpGet("account")]
        public async Task<ActionResult<AccountInfoDTO>> GetAccountInfo()
        {
            var email = User.FindFirstValue(ClaimTypes.Name);
            var accountInfo = await _customerService.GetAccountInfoByEmailAsync(email);
            return Ok(accountInfo);
        }

        [Authorize(Roles = "ADMIN,MANAGER,CLERK")]
        [HttpGet("account/{userId}")]
        public async Task<ActionResult<AccountInfoDTO>> GetAccountInfoByUserId(long userId)
        {
            var accountInfo = await _customerService.GetAccountInfoByUserIdAsync(userId);
            return Ok(accountInfo);
        }

        [Authorize(Roles = "CUSTOMER")]
        [HttpGet("self-kyc")]
        public async Task<ActionResult<KycApplication>> GetOwnKyc()
        {
            var email = User.FindFirstValue(ClaimTypes.Name);
            var kyc = await _customerService.GetKycForCurrentUserAsync(email);
            return Ok(kyc);
        }
    }
}
